Java/src/main/java/com/gildedrose/GildedRose.java

```diff
@@ -1,62 +1,89 @@
-                        if (items[i].sellIn < 11) {
-                            if (items[i].quality < 50) {
-                                items[i].quality = items[i].quality + 1;
-                            }
-                        }
-
-                        if (items[i].sellIn < 6) {
-                            if (items[i].quality < 50) {
-                                items[i].quality = items[i].quality + 1;
-                            }
-                        }
-                    }
-                }
+                dealWithPasses(item);
             }
 
-            if (!items[i].name.equals("Sulfuras, Hand of Ragnaros")) {
-                items[i].sellIn = items[i].sellIn - 1;
+            if (!item.name.equals("Sulfuras, Hand of Ragnaros")) {
+                decrementSellIn(item);
             }
 
-            if (items[i].sellIn < 0) {
-                if (!items[i].name.equals("Aged Brie")) {
-                    if (!items[i].name.equals("Backstage passes to a TAFKAL80ETC concert")) {
-                        if (items[i].quality > 0) {
-                            if (!items[i].name.equals("Sulfuras, Hand of Ragnaros")) {
-                                items[i].quality = items[i].quality - 1;
-                            }
-                        }
-                    } else {
-                        items[i].quality = items[i].quality - items[i].quality;
-                    }
+            if (item.sellIn < 0) {
+                if (!item.name.equals(CHEESE) &&
+                    !item.name.equals(BACKSTAGE_PASSES_TO_A_TAFKAL_80_ETC_CONCERT)) {
+                    dealWithExpiredItem(item);
                 } else {
-                    if (items[i].quality < 50) {
-                        items[i].quality = items[i].quality + 1;
+                    dealWithExpiredCheeseOrPasses(item);
+                }
+            }
+        }
+    }
+
+    private void decrementSellIn(Item item) {
+        item.sellIn = item.sellIn - 1;
+    }
+
+    private void dealWithPasses(Item item) {
+        if (item.quality < 50) {
+            item.quality = item.quality + 1;
+
+            if (item.name.equals(BACKSTAGE_PASSES_TO_A_TAFKAL_80_ETC_CONCERT)) {
+                if (item.sellIn < 11) {
+                    if (item.quality < 50) {
+                        item.quality = item.quality + 1;
                     }
                 }
+
+                // If the sellIn is 5 days or fewer add to the quality again
+                if (item.sellIn < 6) {
+                    item.quality = item.quality + 1;
+                }
             }
         }
     }
-}
\ No newline at end of file
+
+    private void dealWithExpiredItem(Item item) {
+        if (item.quality > 0) {
+            if (!item.name.equals(SULFURAS)) {
+                item.quality = item.quality - 1;
+            }
+        } else {
+            item.quality = 0;
```

Comment:
this else should not be needed as we will only decrease quality if it is larger than 0 so cant go below 0